src/coreclr/src/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs

```diff
@@ -1164,53 +1562,56 @@ internal static bool IsAttributeDefined(RuntimeModule decoratedModule, int decor
             return false;
         }
 
-        private static object[] GetCustomAttributes(
-            RuntimeModule decoratedModule, int decoratedMetadataToken, int pcaCount, RuntimeType? attributeFilterType)
+        private static Attribute? GetCustomAttribute(
+            RuntimeModule decoratedModule, int decoratedMetadataToken, RuntimeType? attributeFilterType)
         {
-            RuntimeType.ListBuilder<object> attributes = default;
-
+            RuntimeType.ListBuilder<Attribute> attributes = default;
             AddCustomAttributes(ref attributes, decoratedModule, decoratedMetadataToken, attributeFilterType, false, default);
 
-            bool useObjectArray = attributeFilterType == null || attributeFilterType.IsValueType || attributeFilterType.ContainsGenericParameters;
-            RuntimeType arrayType = useObjectArray ? (RuntimeType)typeof(object) : attributeFilterType!;
+            if (attributes.Count == 0)
+                return null;
+
+            if (attributes.Count == 1)
+                return attributes[0];
 
-            object[] result = CreateAttributeArrayHelper(arrayType, attributes.Count + pcaCount);
-            for (int i = 0; i < attributes.Count; i++)
-            {
-                result[i] = attributes[i];
-            }
-            return result;
+            throw new AmbiguousMatchException(SR.RFLCT_AmbigCust);
         }
 
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2065:UnrecognizedReflectionPattern",
             Justification = "Linker guarantees presence of all the constructor parameters, property setters and fiels which are accessed by any " +
                             "attribute instantiation which is present in the code linker has analyzed." +
                             "As such the reflection usage in this method should never fail as those methods/fields should always be present.")]
         private static void AddCustomAttributes(
-            ref RuntimeType.ListBuilder<object> attributes,
+            ref RuntimeType.ListBuilder<Attribute> attributes,
             RuntimeModule decoratedModule, int decoratedMetadataToken,
             RuntimeType? attributeFilterType, bool mustBeInheritable,
-            // The derivedAttributes list must be passed by value so that it is not modified with the discovered attributes
-            RuntimeType.ListBuilder<object> derivedAttributes)
+            in RuntimeType.ListBuilder<Attribute> derivedAttributes)
```

Comment:
Does the caller create the copy of the struct manually now? I don't see where it happens. It would be useful to add this comment where it happens.

`in` won't create a copy of the struct, it just assumes that whatever you are using it with is immutable . IMHO, using `in` with mutable structs is very confusing.

Reply:
It's only used in a readonly fashion on indexer to check if the attribute has already been added and is `AllowMultiple` (which is why I changed indexer `get` to be a `readonly` method)

https://github.com/dotnet/runtime/blob/15f7a7950792e8d8a27af2d3f77112cbe27870f6/src/coreclr/src/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs#L1448-L1455

Also copying it won't do much unless it only has one element as it will share the same backing array regardless; unless the linkage breaks because the array gets resized

Reply:
Correction, count will be frozen. Might be easier just to pass in the derived count and use the original `attributes`?

Reply:
Something that makes it easy to see what's going on.

Also - in the existing code, ListBuilder was append only so it was easy to see that copying of the struct gives you immutable snapshot. The ListBuilder is editable now (it has setter). It makes it much harder that the snapshot passed in is immutable.

Reply:
C# gives an error if I try to use readonly on the get but have no setter 😢

![image](https://user-images.githubusercontent.com/1142958/100523234-d69ede00-31a6-11eb-8b21-b80abc8e8cf6.png)

Reply:
Ah, nvm, if only a `get` than it goes on the property
```csharp
public readonly T this[int index]
{
    get
    {
        Debug.Assert(index < Count);
        return (_items != null) ? _items[index] : _item;
    }
}
```

Reply:
Changed